### ISSUE TITLE
FPN-8 | Phase 2 | Train Autoencoder for Unsupervised Anomaly Detection

### ISSUE THREAD

**Issue by LavanyaSrinivas06** [LavanyaSrinivas06/AI-First-Preauth-Fraud-Detection#57]
### Ticket ID: FPN-8  
Phase: Phase 2 — Model Training & Evaluation  
Branch Name: feature/FPN-8-train-autoencoder

---

### Context
An Autoencoder trained on legitimate transactions captures normal behavior and flags deviations as potential frauds.  
This complements the XGBoost model by detecting new or unseen fraud patterns.

---

### Scope
- Implement and train an Autoencoder neural network.  
- Use only legitimate (non-fraud) samples for training.  
- Evaluate performance using reconstruction error, TPR, and FPR.  
- Save trained model and metrics.

---

### Steps

1. **Create training script**
   - File: `scripts/train_autoencoder.py`
   - Load `train.parquet`, filter `Class == 0` (legitimate transactions).
   - Normalize numeric features using same preprocessing pipeline.
   - Build network (Keras or PyTorch):
     ```python
     from tensorflow.keras import layers, models

     input_dim = X_train.shape[1]
     model = models.Sequential([
         layers.Input(shape=(input_dim,)),
         layers.Dense(64, activation='relu'),
         layers.Dense(32, activation='relu'),
         layers.Dense(16, activation='relu'),
         layers.Dense(32, activation='relu'),
         layers.Dense(64, activation='relu'),
         layers.Dense(input_dim, activation='linear')
     ])
     model.compile(optimizer='adam', loss='mse')
     ```

2. **Train model**
   - Use early stopping with validation split (10%).  
   - Epochs: 100, batch size: 512.  
   - Save training history and loss curve.

3. **Evaluate reconstruction error**
   - Compute reconstruction error for test data:
     ```python
     errors = np.mean(np.square(X_test - X_pred), axis=1)
     ```
   - Determine threshold `T` using validation set (e.g., 95th percentile of normal errors).  
   - Compute metrics:
     - True Positive Rate (Recall)  
     - False Positive Rate  
     - ROC-AUC and Precision-Recall AUC.

4. **Persist model and threshold**
   - Save model: `artifacts/autoencoder_model.h5`
   - Save threshold: `artifacts/ae_threshold.txt`
   - Save metrics JSON: `artifacts/ae_metrics.json`

5. **Documentation**
   - Create `docs/autoencoder_training_report.md` including:
     - Architecture summary.
     - Training parameters.
     - Performance metrics.
     - Error distribution plots.

---

### Acceptance Criteria
- [ ] Autoencoder trains successfully on legitimate data.  
- [ ] Reconstruction error threshold determined and documented.  
- [ ] Achieves TPR ≥ 0.85 and FPR ≤ 0.10.  
- [ ] Artifacts (`autoencoder_model.h5`, `ae_threshold.txt`, `ae_metrics.json`) saved.  
- [ ] Report (`autoencoder_training_report.md`) completed with plots and analysis.

---

### Artifacts
- `scripts/train_autoencoder.py`  
- `artifacts/autoencoder_model.h5`  
- `artifacts/ae_threshold.txt`  
- `artifacts/ae_metrics.json`  
- `docs/autoencoder_training_report.md`

---

### Helpful Links
- Keras Autoencoder tutorial: https://keras.io/examples/structured_data/autoencoder/  
- PyTorch Autoencoder reference: https://pytorch.org/tutorials/  
- Precision-Recall evaluation: https://scikit-learn.org/stable/auto_examples/model_selection/plot_precision_recall.html

---

### Effort Estimate
M (1–2 days)
